File: mods/java/fabric/src/main/java/com/alaydriem/bedrockvoicechat/fabric/FabricConfigProvider.java
package com.alaydriem.bedrockvoicechat.fabric;

import com.alaydriem.bedrockvoicechat.api.ConfigProvider;
import com.alaydriem.bedrockvoicechat.config.ModConfig;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import net.fabricmc.loader.api.FabricLoader;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Reader;
import java.io.Writer;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Fabric-specific configuration provider using JSON files.
 */
public class FabricConfigProvider implements ConfigProvider {
    private static final Logger LOGGER = LoggerFactory.getLogger("Bedrock Voice Chat");
    private static final Path CONFIG_PATH =
            FabricLoader.getInstance().getConfigDir().resolve("bedrock-voice-chat.json");
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();

    @Override
    public ModConfig load() {
        if (!Files.exists(CONFIG_PATH)) {
            LOGGER.warn("Config not found, creating default config");
            createDefaultIfMissing();
            return new ModConfig();
        }

        try (Reader reader = Files.newBufferedReader(CONFIG_PATH)) {
            return GSON.fromJson(reader, ModConfig.class);
        } catch (Exception e) {
            LOGGER.error("Failed to load config", e);
            return new ModConfig();
        }
    }

    @Override
    public void save(ModConfig config) {
        try (Writer writer = Files.newBufferedWriter(CONFIG_PATH)) {
            GSON.toJson(config, writer);
        } catch (Exception e) {
            LOGGER.error("Failed to save config", e);
        }
    }

    @Override
    public void createDefaultIfMissing() {
        if (Files.exists(CONFIG_PATH)) {
            return;
        }

        ModConfig defaultConfig = new ModConfig();
        defaultConfig.setBvcServer("");
        defaultConfig.setAccessToken("");
        defaultConfig.setMinimumPlayers(2);

        try (Writer writer = Files.newBufferedWriter(CONFIG_PATH)) {
            GSON.toJson(defaultConfig, writer);
            LOGGER.info("Default config created at: {}", CONFIG_PATH);
        } catch (Exception e) {
            LOGGER.error("Failed to create default config", e);
        }
    }
}

File: mods/java/fabric/src/main/kotlin/com/alaydriem/bedrockvoicechat/fabric/FabricConfigProvider.kt
package com.alaydriem.bedrockvoicechat.fabric

import com.alaydriem.bedrockvoicechat.api.ConfigProvider
import com.alaydriem.bedrockvoicechat.config.ModConfig
import com.google.gson.GsonBuilder
import net.fabricmc.loader.api.FabricLoader
import org.slf4j.LoggerFactory
import java.nio.file.Files

/**
 * Fabric-specific configuration provider using JSON files.
 */
class FabricConfigProvider : ConfigProvider {
    companion object {
        private val LOGGER = LoggerFactory.getLogger("Bedrock Voice Chat")
        private val CONFIG_PATH = FabricLoader.getInstance().configDir.resolve("bedrock-voice-chat.json")
        private val GSON = GsonBuilder().setPrettyPrinting().create()
    }

    override fun load(): ModConfig {
        LOGGER.debug("Looking for config at: {}", CONFIG_PATH.toAbsolutePath())

        if (!Files.exists(CONFIG_PATH)) {
            LOGGER.warn("Config not found at {}, creating default config", CONFIG_PATH.toAbsolutePath())
            createDefaultIfMissing()
            return ModConfig()
        }

        return try {
            LOGGER.debug("Loading config from: {}", CONFIG_PATH.toAbsolutePath())
            val config = Files.newBufferedReader(CONFIG_PATH).use { reader ->
                GSON.fromJson(reader, ModConfig::class.java)
            }
            LOGGER.debug("Loaded config - bvcServer: {}, accessToken: {}, minimumPlayers: {}",
                config.bvcServer?.take(20) ?: "null",
                if (config.accessToken.isNullOrBlank()) "null/blank" else "***set***",
                config.minimumPlayers)
            config
        } catch (e: Exception) {
            LOGGER.error("Failed to load config from {}", CONFIG_PATH.toAbsolutePath(), e)
            ModConfig()
        }
    }

    override fun save(config: ModConfig) {
        try {
            Files.newBufferedWriter(CONFIG_PATH).use { writer ->
                GSON.toJson(config, writer)
            }
        } catch (e: Exception) {
            LOGGER.error("Failed to save config", e)
        }
    }

    override fun createDefaultIfMissing() {
        if (Files.exists(CONFIG_PATH)) {
            return
        }

        val defaultConfig = ModConfig().apply {
            bvcServer = ""
            accessToken = ""
            minimumPlayers = 2
        }

        try {
            Files.newBufferedWriter(CONFIG_PATH).use { writer ->
                GSON.toJson(defaultConfig, writer)
            }
            LOGGER.info("Default config created at: {}", CONFIG_PATH)
        } catch (e: Exception) {
            LOGGER.error("Failed to create default config", e)
        }
    }
}
